vc-got.el

```diff
@@ -605,31 +605,27 @@ Value is returned as floating point fractional number of days."
 
 (defun vc-got-previous-revision (file rev)
   "Return the revision number that precedes REV for FILE, or nil if no such revision exists."
-  (let (process-file-side-effects)
-    (vc-got-with-worktree file
-      (with-temp-buffer
-        (when (zerop (apply #'vc-got--call "log" (list "-c" rev "-l" "2" "-R" file)))
-          (keep-lines "^commit")
-          (when (looking-at "^commit \\([a-z0-9]+\\)")
-            (match-string-no-properties 1)))))))
+  (with-temp-buffer
+    (vc-got--log file 2 rev nil nil t)
+    (keep-lines "^commit")
```

Comment:
keep-lines doesn't do anything because the point is at the end of the buffer after vc-got--log.  (process-file leaves the point at the end of the output).  looking-at fails and so nil is returned.

keep-lines changes only the line past the point, so you either `(goto-char (point-min))` before keep-lines or call keep-lines with a rstart and rend argument.

Even with this I get an "Buffer \*Annotate vc-got.el (rev xyz)\* is not associated with a file" though.

Reply:
Ah, I tested with empty buffer and run C-u M-! to run got log command and that left the point at the start of buffer. I had the goto-char before so I'll restore that there.

About the error I think the vc-got-annotate-extract-revision-at-line should return (cons revision . filename) instead of just revision. That at least the what the vc-annotate-extract-revision-at-line doc string says but there's different comment on vc.el. There its said "If the backend supports annotating through copies and renames, and displays a file name and a revision, then return a cons (REVISION . FILENAME)." I don't know what it tries to say exactly but I think its worth to try to add the filename as its return value.

Reply:
I'm not sure it's related to the error, but yeah, annotate-extract-revision-at-line should return `(revision . filename)`.  Unlike CVS, git (and thus got) supports renaming files and probably helps when the previous/next revision points to a file with a different name.  (to be fair, from a quick look only vc-git seems to return a cons, vc-svn and vc-hg don't)

I'm not sure how to get the filename though.  One idea could be define a buffer-local variable and set it in vc-got-annotate-command.  Something along the lines of

```elisp
(defvar-local vc-got--annotate-file nil
  "The file being annotated")

(defun vc-got-annotate-command (file buf &optional rev)
  "..."
  (with-current-buffer buf
    (setq vc-got--annotate-file file)
    ...))

(defun vc-got-annotate-extract-revision-at-line ()
  ...
  (cons (match-string-no-properties 1) vc-got--annotate-file))
```

but it seems pretty hackish.  (Also, FYI `with-current-buffer` is equivalent to `(save-current-buffer (set-buffer buf) ...)`)

Maybe we can ask the devs of got to include the filename in the output of blame?  Not sure if it's a good approach though.

Reply:
It would seem a bit redundant to print filename in blame, you give the path as argument to blame so the caller should know it.

I don't think setting local variable would work, what if user lists annotation for another file between toggleing prev/next in old file. It would list wrong files contents.
How about this, parse the filename from the buffer name?

```
(defun vc-got-annotate-extract-revision-at-line ()
  "Returns revision corresponding to the current line or nil."
  (save-excursion
    (beginning-of-line)
    (when (string-match "*Annotate \\(.+\\) (rev [a-z0-9]+)\\*$" (buffer-name))
      (let ((fname (match-string-no-properties 1)))
        (when (looking-at vc-got-annotate-re)
          (cons (match-string-no-properties 1)
                (expand-file-name fname (vc-got-root default-directory))))))))
```

Reply:
There's a problem with parsing the buffer name.  The buffer name contains only the filename, not the path.  Just tried on a project: visited `src/clj/foo/search.clj` and the annotate buffer (from vc-git) is called "\*Annotate search.clj\*".  If I have another `search.clj` in that project, say `test/foo/search.clj` you don't know which of those two is the one I'm interested in.

An aside, sometimes I got confused by the buffer-local variables, but I don't think that your example is right.  if I run vc-annotate on file1 I get a buffer "Annotate file1".  Then I run vc-annotate on file2, and I get a buffer "Annotate file2".   Now, vc-got--annotate-file is "file1" in "Annotate file1" and "file2" in buffer "Annotate file2", since it's buffer local.  it's bound to that buffer, regardless if I change it on another IIUC.

Anyway, all of this is meaningless.  I just remembered reading about a `(vc-deduce-fileset)` while reading some diff-related code.  It seems that (at least for vc-git, I don't have the time to further test right now) it yields the full path to the file if called in an annotation buffer.  Try it with M-: (vc-deduce-fileset) on a annotation buffer.

Reply:
nevermind, `(vc-deduce-fileset)` doesn't seem to work.  But at least we can access the original file through the buffer-local `vc-annotate-parent-file`: vc does the hack for us!  (now the question is: given that c-annotate-parent-file` is provided by VC, why it requires us to return `(rev . file)` when it already knows the filename? )

Anyway, I'm curios why we have cases where VC errors about an annotated buffer that isn't associated with a file.

Reply:
(sorry for the spam)

Eureka!  I think I got why `annotate-extract-revision-at-line` may return `(rev . file)`.  The problem is not getting the original file name, in fact vc does that by itself (in `vc-annotate` it set a local variable for that), but to handle the case when file gets renamed.  It makes sense for got (or any other vcs really) to output the filename in the blame output because a line in `file1` may come from `file2`, if at some point in the history `file2` was renamed to `file1`.

Anyway, this shouldn't block this PR.  For the time being we can simply return the revision, as you're doing.  If you don't beat me to it, between this night and tomorrow morning I should find the time to debug the remaining issues in the next/prev revision functions.

Reply:
Got previous to semi-working by adding (setq buffer-file-name file) to vc-got-annotate-command and then returning it from annotate-extract-revision-at-line. Though it messed the buffer name for some reason. It started to loop "Annotate Annotate Annotate .init.el ...".

Reply:
Yeah, I spent some time debugging this and part of the fault is due to `vc-ensure-vc-buffer` not recognising the annotate buffer as managed by got.  This is probably my fault for not using `vc-do-command`, I still have to investigate that.

I managed to get next and previous revision working, I'll add comments to indicate the modification I had to do.